Add WASM opcode counter

hyjorc1 opened the pull request:
#### 70ccbd9daaa2db326dfed9a1d84f7be4ffdddaa7
<pre>
Add WASM opcode counter
<a href="https://bugs.webkit.org/show_bug.cgi?id=250213">https://bugs.webkit.org/show_bug.cgi?id=250213</a>
rdar://103956646

Reviewed by Justin Michaud.

This patch includes three things:

a) Added `atob` and `btoa` to jsc shell as helpers for loading tfjs pre-trained models.

b) Skipped JetStream3 tests UniPoker and bigint-paillier, since UniPoker
   fails on browser with error messages and bigint-paillier hangs with
   `deterministicRandom` is true.

c) Added a WASM opcode counter. Use `dumpWasmOpcodeStatistics=1` to enable
   opcode collection. And use `notifyutil -v -p com.apple.WebKit.wasm.op.stat`
   to dump statistics sorted by counts in descending order for all WASM opcodes.
   An example of dumped statistics:

      &lt;WASM.OP.STAT&gt;&lt;92428&gt; Registering callback for wasm opcode statistics.
      &lt;WASM.OP.STAT&gt;&lt;92428&gt; Use `notifyutil -v -p com.apple.WebKit.wasm.op.stat` to dump statistics.
      &lt;WASM.EXT.SIMD.OP.STAT&gt;&lt;92428&gt; wasm extended SIMD opcode use coverage 32.63%.
      &lt;WASM.EXT.SIMD.OP.STAT&gt;&lt;92428&gt;    V128Load: 4588
      &lt;WASM.EXT.SIMD.OP.STAT&gt;&lt;92428&gt;    F32x4Add: 2522
      ...
      &lt;WASM.EXT.ATOMIC.OP.STAT&gt;&lt;92428&gt; wasm extended atomic opcode use coverage 0.00%.
      &lt;WASM.EXT.ATOMIC.OP.STAT&gt;&lt;92428&gt;    MemoryAtomicNotify: 0
      &lt;WASM.EXT.ATOMIC.OP.STAT&gt;&lt;92428&gt;    MemoryAtomicWait32: 0
      ...
      &lt;WASM.GC.OP.STAT&gt;&lt;92428&gt; wasm GC opcode use coverage 0.00%.
      &lt;WASM.GC.OP.STAT&gt;&lt;92428&gt;    StructNewCanon: 0
      &lt;WASM.GC.OP.STAT&gt;&lt;92428&gt;    StructGet: 0
      ...
      &lt;WASM.BASE.OP.STAT&gt;&lt;92428&gt; wasm base opcode use coverage 74.49%.
      &lt;WASM.BASE.OP.STAT&gt;&lt;92428&gt;    GetLocal: 80967
      &lt;WASM.BASE.OP.STAT&gt;&lt;92428&gt;    I32Const: 28805
      ...

* JSTests/stress/atob-btoa.js: Added.
(assert):
(shouldThrow):
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/jsc.cpp:
(JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseBody):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseExpression):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseUnreachableExpression):
* Source/JavaScriptCore/wasm/WasmOpcodeCounter.cpp: Added.
(JSC::Wasm::WasmOpcodeCounter::singleton):
(JSC::Wasm::WasmOpcodeCounter::dump):
(JSC::Wasm::WasmOpcodeCounter::registerDispatch):
(JSC::Wasm::WasmOpcodeCounter::increment):
* Source/JavaScriptCore/wasm/WasmOpcodeCounter.h: Added.
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
(JSC::Wasm::countNumberOfWasmExtendedSIMDOpcodes):
(JSC::Wasm::isRegisteredWasmExtendedSIMDOpcode):
(JSC::Wasm::dumpExtSIMDOpType):
(JSC::Wasm::countNumberOfWasmExtendedAtomicOpcodes):
(JSC::Wasm::isRegisteredExtenedAtomicOpcode):
(JSC::Wasm::dumpExtAtomicOpType):
(JSC::Wasm::countNumberOfWasmGCOpcodes):
(JSC::Wasm::isRegisteredGCOpcode):
(JSC::Wasm::dumpGCOpType):
(JSC::Wasm::countNumberOfWasmBaseOpcodes):
(JSC::Wasm::isRegisteredBaseOpcode):
(JSC::Wasm::dumpOpType):

Canonical link: <a href="https://commits.webkit.org/258678@main">https://commits.webkit.org/258678@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f8f19533a3d875846b41cd86e8e9b44c4824e54c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/102607 "8 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/11738 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/35642 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/111876 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/172098 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/12743 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/2643 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/94884 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/109579 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/108385 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/9771 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92990 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/37429 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/91625 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/24495 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79174 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/92846 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5195 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/25914 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/89169 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/2872 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/5354 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2366 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29629 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/11376 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45403 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/92094 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5961 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7086 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20631 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/wasm/WasmFunctionParser.h:
```diff
@@ -608,6 +609,9 @@ template<typename Context>
 template<bool isReachable, typename>
 auto FunctionParser<Context>::simd(SIMDLaneOperation op, SIMDLane lane, SIMDSignMode signMode, B3::Air::Arg optionalRelation) -> PartialResult
 {
+    if (Options::dumpWasmSimdOperationStatistics())
```
Use `UNLIKELY`

MenloDorian commented on Source/JavaScriptCore/Sources.txt:
```diff
@@ -1136,6 +1136,7 @@ wasm/WasmOperations.cpp
 wasm/WasmPlan.cpp
 wasm/WasmSectionParser.cpp
 wasm/WasmTypeDefinition.cpp
+wasm/WasmSIMDLaneOperationCounter.cpp
```
Why only Wasm SIMD Lane operations?  Why not all Wasm opcode?

MenloDorian commented on Source/JavaScriptCore/wasm/WasmSIMDLaneOperationCounter.h:
```diff
@@ -0,0 +1,51 @@
+public:
+    static WasmSIMDLaneOperationCounter& singleton();
+
+    void registerAndIncrement(SIMDLaneOperation);
```
nit: I think you don't have to spell out that this function also initializes the counters.  Let's call this `incrementSIMDOpcode` instead.  The rest is an implementation detail.

Also add an `incrementBaseOpcode` for Wasm base opcodes.  I don't know if "base" opcode is an appropriate term to use.  Maybe Justin can give you better names for these.

MenloDorian commented on Source/JavaScriptCore/jsc.cpp:
```diff
@@ -1379,6 +1385,39 @@ JSC_DEFINE_HOST_FUNCTION(functionPrettyPrint, (JSGlobalObject* globalObject, Cal
+    String stringToEncode = string->value(globalObject);
+
+    if (stringToEncode.isNull())
+        return JSValue::encode(jsString(vm, String()));
```
I think you want `jsEmptyString(vm)` instead of `jsString(vm, String())`.  BTW, `String()` is the NULL string as opposed to the empty string.  However, `jsString()` converts it to the empty string.  Using `jsEmptyString(vm)` directly cuts out a lot of unnecessary work in between.

MenloDorian commented on Source/JavaScriptCore/wasm/WasmSIMDLaneOperationCounter.h:
```diff
@@ -0,0 +1,51 @@
+
+namespace JSC {
+
+class WasmSIMDLaneOperationCounter {
```
I suggest expanding this to cover all Wasm opcodes.  Rename this to `WasmOpcodeCounter`.

MenloDorian commented on Source/JavaScriptCore/wasm/WasmSIMDLaneOperationCounter.h:
```diff
@@ -0,0 +1,51 @@
+
+private:
+    constexpr static unsigned m_totalOperationCount = countNumberOfSIMDLaneOperations();
+    Atomic<uint64_t> m_counter[m_totalOperationCount];
```
Add a set for Wasm base opcodes, and have `incrementBaseOpcode` increment those.

MenloDorian commented on Source/JavaScriptCore/wasm/WasmSIMDLaneOperationCounter.h:
```diff
@@ -0,0 +1,51 @@
+
+    void registerAndIncrement(SIMDLaneOperation);
+
+    void dump();
```
Change the dumper to dump both the base and SIMD opcode counts.

MenloDorian reviewed (changes requested):
Looks good but I we should also count the base Wasm opcodes.

justinmichaud reviewed (approved):
r=me

hyjorc1 committed:
Add WASM opcode counter
https://bugs.webkit.org/show_bug.cgi?id=250213
rdar://103956646

Reviewed by Justin Michaud.

This patch includes three things:

a) Added `atob` and `btoa` to jsc shell as helpers for loading tfjs pre-trained models.

b) Skipped JetStream3 tests UniPoker and bigint-paillier, since UniPoker
   fails on browser with error messages and bigint-paillier hangs with
   `deterministicRandom` is true.

c) Added a WASM opcode counter. Use `dumpWasmOpcodeStatistics=1` to enable
   opcode collection. And use `notifyutil -v -p com.apple.WebKit.wasm.op.stat`
   to dump statistics sorted by counts in descending order for all WASM opcodes.
   An example of dumped statistics:

      <WASM.OP.STAT><92428> Registering callback for wasm opcode statistics.
      <WASM.OP.STAT><92428> Use `notifyutil -v -p com.apple.WebKit.wasm.op.stat` to dump statistics.
      <WASM.EXT.SIMD.OP.STAT><92428> wasm extended SIMD opcode use coverage 32.63%.
      <WASM.EXT.SIMD.OP.STAT><92428>    V128Load: 4588
      <WASM.EXT.SIMD.OP.STAT><92428>    F32x4Add: 2522
      ...
      <WASM.EXT.ATOMIC.OP.STAT><92428> wasm extended atomic opcode use coverage 0.00%.
      <WASM.EXT.ATOMIC.OP.STAT><92428>    MemoryAtomicNotify: 0
      <WASM.EXT.ATOMIC.OP.STAT><92428>    MemoryAtomicWait32: 0
      ...
      <WASM.GC.OP.STAT><92428> wasm GC opcode use coverage 0.00%.
      <WASM.GC.OP.STAT><92428>    StructNewCanon: 0
      <WASM.GC.OP.STAT><92428>    StructGet: 0
      ...
      <WASM.BASE.OP.STAT><92428> wasm base opcode use coverage 74.49%.
      <WASM.BASE.OP.STAT><92428>    GetLocal: 80967
      <WASM.BASE.OP.STAT><92428>    I32Const: 28805
      ...

* JSTests/stress/atob-btoa.js: Added.
(assert):
(shouldThrow):
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/jsc.cpp:
(JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/OptionsList.h:
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser<Context>::parseBody):
(JSC::Wasm::FunctionParser<Context>::parseExpression):
(JSC::Wasm::FunctionParser<Context>::parseUnreachableExpression):
* Source/JavaScriptCore/wasm/WasmOpcodeCounter.cpp: Added.
(JSC::Wasm::WasmOpcodeCounter::singleton):
(JSC::Wasm::WasmOpcodeCounter::dump):
(JSC::Wasm::WasmOpcodeCounter::registerDispatch):
(JSC::Wasm::WasmOpcodeCounter::increment):
* Source/JavaScriptCore/wasm/WasmOpcodeCounter.h: Added.
* Source/JavaScriptCore/wasm/WasmTypeDefinition.h:
(JSC::Wasm::countNumberOfWasmExtendedSIMDOpcodes):
(JSC::Wasm::isRegisteredWasmExtendedSIMDOpcode):
(JSC::Wasm::dumpExtSIMDOpType):
(JSC::Wasm::countNumberOfWasmExtendedAtomicOpcodes):
(JSC::Wasm::isRegisteredExtenedAtomicOpcode):
(JSC::Wasm::dumpExtAtomicOpType):
(JSC::Wasm::countNumberOfWasmGCOpcodes):
(JSC::Wasm::isRegisteredGCOpcode):
(JSC::Wasm::dumpGCOpType):
(JSC::Wasm::countNumberOfWasmBaseOpcodes):
(JSC::Wasm::isRegisteredBaseOpcode):
(JSC::Wasm::dumpOpType):

Canonical link: https://commits.webkit.org/258678@main
